tests/0143-exponential_backoff.c

```diff
@@ -0,0 +1,418 @@
+        for(size_t i = 0; (i < request_cnt) && (retry_count < num_retries); i++){
+                TEST_SAY("Broker Id : %d API Key : %d Timestamp : %lld\n",rd_kafka_mock_request_id(requests[i]), rd_kafka_mock_request_api_key(requests[i]), rd_kafka_mock_request_timestamp(requests[i]));
+                if(rd_kafka_mock_request_api_key(requests[i]) == RD_KAFKAP_FindCoordinator){
+                        if(previous_request_ts != -1){
+                                int64_t time_difference = (rd_kafka_mock_request_timestamp(requests[i]) - previous_request_ts)/1000;
+                                TEST_ASSERT(((time_difference > high - buffer) && (time_difference < high + buffer)) || ((time_difference > low - buffer) && (time_difference < low + buffer)) ,"Time difference should be close to 1 or 2 seconds, it is %lld ms instead.\n",time_difference);
+                                retry_count++;
+                        }
+                        previous_request_ts = rd_kafka_mock_request_timestamp(requests[i]);
+                }
+        }
+        rd_kafka_destroy(consumer);
+        rd_kafka_mock_clear_requests(mcluster);
+        SUB_TEST_PASS();
+}
+
+/** OffsetCommit Test
+ * We fail the request with RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS so that
+ * the request is retried with the exponential backoff. The max retries allowed is 2.
+*/
+void test_OffsetCommit(rd_kafka_mock_cluster_t *mcluster,const char *topic, rd_kafka_conf_t *conf){
+        rd_kafka_mock_request_t **requests = NULL;
+        size_t request_cnt = 0;
+        int64_t previous_request_ts = -1;
+        int32_t retry_count = 0;
+        rd_kafka_t *consumer;
+
+        SUB_TEST();
+        test_conf_set(conf, "topic.metadata.refresh.interval.ms","-1");
+        test_conf_set(conf, "debug" ,"generic,cgrp");
+        test_conf_set(conf, "auto.offset.reset", "earliest");
+        test_conf_set(conf, "enable.auto.commit", "false");
+
+        consumer = test_create_consumer(topic, NULL, rd_kafka_conf_dup(conf), NULL);
+        test_consumer_subscribe(consumer,topic);
+        rd_kafka_message_t *rkm = rd_kafka_consumer_poll(consumer,10*1000);
+        if (rkm)
+                rd_kafka_message_destroy(rkm);
+        rd_sleep(4);
+        rd_kafka_mock_push_request_errors(mcluster, RD_KAFKAP_OffsetCommit, 2,
+                                            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+                                            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS);
+
+        rd_kafka_topic_partition_list_t *offsets = rd_kafka_topic_partition_list_new(1);
+        rd_kafka_topic_partition_t *rktpar = rd_kafka_topic_partition_list_add(offsets,topic,0);
+        rktpar->offset = 4;
+        rd_kafka_commit(consumer,offsets,0);
+        rd_kafka_topic_partition_list_destroy(offsets);
+        rd_sleep(3);
+        requests = rd_kafka_mock_get_requests(mcluster,&request_cnt);
+        for(size_t i = 0; i < request_cnt; i++){
```

Comment:
It has different body for all the functions and difference is not subtle to make it into a new function.

Reply:
Please don't resolve comments for which you are starting a discussion, it makes it difficult to address the point you've raised.

I don't agree with this, because in multiple test cases, code for checking is copy pasted.
In particular:
> You can make multiple function, for some of the cases where handling differently is needed

The checking code for Produce, OffsetCommit, and Produce Fast Leader query can be made into one function.
The checking code for JoinGroup and Heartbeat plus FindCoordinator can also be made into one function.

That only leaves Fetch+Fast leader query and FindCoordinator can remain as separate loops

Let me know if you have any difficulty doing this, I can help.